File: src/commands/calendar.ts
import { formatTimestamp } from "../lib/utils.ts";
import { Command } from "commander";
import { getEnrolledCoursesApi, getCalendarEventsApi } from "../lib/moodle.ts";
import { createApiContext } from "../lib/auth.ts";
import fs from "node:fs";

export function registerCalendarCommand(program: Command): void {
  const calendarCmd = program.command("calendar");
  calendarCmd.description("Calendar operations");

  calendarCmd
    .command("events")
    .description("List calendar events")
    .option("--upcoming", "Show only upcoming events")
    .option("--days <n>", "Number of days ahead to look", "30")
    .option("--course <id>", "Filter by course ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (options, command) => {
      const days = parseInt(options.days, 10);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const courses = await getEnrolledCoursesApi(apiContext.session);

      const now = Math.floor(Date.now() / 1000);
      const endTime = now + (days * 24 * 60 * 60);

      let allEvents = [];

      if (options.course) {
        const courseId = parseInt(options.course, 10);
        const events = await getCalendarEventsApi(apiContext.session, {
          startTime: now,
          endTime: endTime,
        });
        allEvents = events.filter(e => e.courseid === courseId);
      } else {
        const results = await Promise.allSettled(
          courses.map(course =>
            getCalendarEventsApi(apiContext.session, {
              courseId: course.id,
              startTime: now,
              endTime: endTime,
            })
          )
        );
        for (const result of results) {
          if (result.status === "fulfilled") allEvents.push(...result.value);
        }
      }

      allEvents.sort((a, b) => a.timestart - b.timestart);

      let filteredEvents = allEvents;
      if (options.upcoming) {
        filteredEvents = allEvents.filter(e => e.timestart > now);
      }

      console.log(JSON.stringify({
        status: "success",
        timestamp: new Date().toISOString(),
        total_events: allEvents.length,
        upcoming: allEvents.filter(e => e.timestart > now).length,
        by_type: allEvents.reduce((acc, e) => {
          acc[e.eventtype] = (acc[e.eventtype] || 0) + 1;
          return acc;
        }, {} as Record<string, number>),
      }));
      for (const e of filteredEvents) {
        console.log(JSON.stringify({
          id: e.id,
          name: e.name,
          description: e.description,
          course_id: e.courseid,
          event_type: e.eventtype,
          start_time: formatTimestamp(e.timestart),
          end_time: e.timeduration ? formatTimestamp(e.timestart + Math.floor(e.timeduration / 1000)) : null,
          location: e.location,
        }));
      }
    });

  calendarCmd
    .command("export")
    .description("Export calendar events to file")
    .option("--output <path>", "Output file path", "./calendar.json")
    .option("--days <n>", "Number of days ahead to include", "30")
    .action(async (options, command) => {
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const courses = await getEnrolledCoursesApi(apiContext.session);

      const now = Math.floor(Date.now() / 1000);
      const days = parseInt(options.days, 10);
      const endTime = now + (days * 24 * 60 * 60);

      const allEvents = [];

      const results = await Promise.allSettled(
        courses.map(course =>
          getCalendarEventsApi(apiContext.session, {
            courseId: course.id,
            startTime: now,
            endTime: endTime,
          })
        )
      );
      for (const result of results) {
        if (result.status === "fulfilled") allEvents.push(...result.value);
      }

      allEvents.sort((a, b) => a.timestart - b.timestart);

      const exportData = {
        exported_at: new Date().toISOString(),
        time_range: {
          start: new Date(now * 1000).toISOString(),
          end: new Date(endTime * 1000).toISOString(),
          days: days,
        },
        events: allEvents.map(e => ({
          id: e.id,
          name: e.name,
          description: e.description,
          course_id: e.courseid,
          event_type: e.eventtype,
          start_time: formatTimestamp(e.timestart),
          end_time: e.timeduration ? formatTimestamp(e.timestart + Math.floor(e.timeduration / 1000)) : null,
          location: e.location,
        })),
        summary: {
          total_events: allEvents.length,
          by_type: allEvents.reduce((acc, e) => {
            acc[e.eventtype] = (acc[e.eventtype] || 0) + 1;
            return acc;
          }, {} as Record<string, number>),
        },
      };

      fs.writeFileSync(options.output, JSON.stringify(exportData));

      apiContext.log.success(`Exported ${allEvents.length} events to ${options.output}`);
    });
}

File: src/commands/calendar.rs
use anyhow::Result;
use crate::Cli;
use crate::moodle::course::get_enrolled_courses_api;
use crate::moodle::calendar::get_calendar_events_api;
use crate::output::format_and_output;
use crate::utils::format_moodle_date;
use std::fs;
use super::ApiCtx;

pub async fn run(cmd: &crate::CalendarCommands, cli: &Cli) -> Result<()> {
    let ctx = ApiCtx::build(cli)?;

    match cmd {
        crate::CalendarCommands::Events { upcoming, days, course } => {
            let now = chrono::Utc::now().timestamp();
            let end_time = now + (*days as i64 * 86400);

            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "inprogress").await?;

            let mut all_events = Vec::new();

            if let Some(course_id) = course {
                let events = get_calendar_events_api(
                    &ctx.client, &ctx.session,
                    Some(*course_id), Some(now), Some(end_time),
                ).await?;
                all_events.extend(events.into_iter().filter(|e| e.courseid == Some(*course_id)));
            } else {
                for c in &courses {
                    match get_calendar_events_api(
                        &ctx.client, &ctx.session,
                        Some(c.id), Some(now), Some(end_time),
                    ).await {
                        Ok(events) => all_events.extend(events),
                        Err(e) => ctx.log.warn(&format!("Failed to get events for course {}: {}", c.id, e)),
                    }
                }
            }

            all_events.sort_by_key(|e| e.timestart);

            let filtered: Vec<_> = if *upcoming {
                all_events.iter().filter(|e| e.timestart > now).collect()
            } else {
                all_events.iter().collect()
            };

            let items: Vec<serde_json::Value> = filtered.iter().map(|e| serde_json::json!({
                "id": e.id,
                "name": e.name,
                "description": e.description,
                "course_id": e.courseid,
                "event_type": e.eventtype,
                "start_time": format_moodle_date(Some(e.timestart)),
                "end_time": e.timeduration.map(|d| format_moodle_date(Some(e.timestart + d))),
                "location": e.location,
            })).collect();

            let upcoming_count = all_events.iter().filter(|e| e.timestart > now).count();
            ctx.log.info(&format!(
                "Total: {} events, {} upcoming",
                all_events.len(), upcoming_count
            ));

            format_and_output(&items, ctx.output, None);
        }

        crate::CalendarCommands::Export { output, days } => {
            let now = chrono::Utc::now().timestamp();
            let end_time = now + (*days as i64 * 86400);

            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "inprogress").await?;
            let mut all_events = Vec::new();

            for c in &courses {
                match get_calendar_events_api(
                    &ctx.client, &ctx.session,
                    Some(c.id), Some(now), Some(end_time),
                ).await {
                    Ok(events) => all_events.extend(events),
                    Err(e) => ctx.log.warn(&format!("Failed to get events for course {}: {}", c.id, e)),
                }
            }

            all_events.sort_by_key(|e| e.timestart);

            let mut by_type = serde_json::Map::new();
            for event in &all_events {
                let count = by_type.entry(event.eventtype.clone())
                    .or_insert(serde_json::json!(0));
                *count = serde_json::json!(count.as_u64().unwrap_or(0) + 1);
            }

            let export_data = serde_json::json!({
                "exported_at": chrono::Utc::now().to_rfc3339(),
                "time_range": {
                    "start": chrono::DateTime::from_timestamp(now, 0)
                        .map(|t| t.to_rfc3339())
                        .unwrap_or_default(),
                    "end": chrono::DateTime::from_timestamp(end_time, 0)
                        .map(|t| t.to_rfc3339())
                        .unwrap_or_default(),
                    "days": days,
                },
                "events": all_events.iter().map(|e| serde_json::json!({
                    "id": e.id,
                    "name": e.name,
                    "description": e.description,
                    "course_id": e.courseid,
                    "event_type": e.eventtype,
                    "start_time": format_moodle_date(Some(e.timestart)),
                    "end_time": e.timeduration.map(|d| format_moodle_date(Some(e.timestart + d))),
                    "location": e.location,
                })).collect::<Vec<_>>(),
                "summary": {
                    "total_events": all_events.len(),
                    "by_type": serde_json::Value::Object(by_type),
                },
            });

            let json_str = serde_json::to_string_pretty(&export_data)?;
            if let Some(parent) = output.parent() {
                fs::create_dir_all(parent)?;
            }
            fs::write(output, &json_str)?;
            ctx.log.success(&format!("Exported {} events to {}", all_events.len(), output.display()));
        }
    }

    Ok(())
}
